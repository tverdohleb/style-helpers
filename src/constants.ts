enum Metric {
  EM = 'em',
  PERCENTS = '%',
  PX = 'px',
  REM = 'rem',
}

export { Metric };
