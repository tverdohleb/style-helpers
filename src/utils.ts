import { Metric } from './constants';
import tinycolor from 'tinycolor2';

/**
 * Wrap value to string using metric specified
 */
const createMetricString = (value: number, metric: Metric): string => `${value}${metric} `;

/**
 * Wrap value to rem string
 */
const rem = (value: number, base: number = 16): string => createMetricString(value / base, Metric.REM);

/**
 * Wrap value to px string
 */
const px = (value: number): string => createMetricString(value, Metric.PX);

/**
 * Wrap value to em string
 */
const em = (value: number): string => createMetricString(value, Metric.EM);

/**
 * Wrap value to % string
 */
const percent = (value: number): string => createMetricString(value, Metric.PERCENTS);

/**
 * Convert color and wrap to rgba string
 */
const rgba = (color: tinycolor.ColorInput, opacity: number): string => {
  const { r, g, b } = tinycolor(color).toRgb();

  return `rgba(${r}, ${g}, ${b}, ${opacity})`;
};

export { rem, px, percent, em, createMetricString, rgba };
