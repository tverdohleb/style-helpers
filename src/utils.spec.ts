import { createMetricString, rem, px, percent, em, rgba } from './utils';
import { Metric } from './constants';

describe('Generate metric strings', () => {
  it('Should return correct metric string', () => {
    const value = 5;

    const result = createMetricString(value, Metric.EM);

    expect(result).toEqual(`${value}${Metric.EM} `);
  });

  it('Should return EM string', () => {
    const value = 5;

    const result = em(value);

    expect(result).toEqual(`${value}${Metric.EM} `);
  });

  it('Should return correct PX string', () => {
    const value = 5;

    const result = px(value);

    expect(result).toEqual(`${value}${Metric.PX} `);
  });

  it('Should return correct % string', () => {
    const value = 5;

    const result = percent(value);

    expect(result).toEqual(`${value}${Metric.PERCENTS} `);
  });

  it('Should return correct REM string', () => {
    const value = 16;
    const base = 16;

    const result = rem(value, base);

    expect(result).toEqual(`${value / base}${Metric.REM} `);
  });

  it('Should return correct REM string with default base', () => {
    const value = 16;

    const result = rem(value);

    expect(result).toMatchSnapshot();
  });

  it('Should return correct RGBA string for hex color', () => {
    const result = rgba('#ABCDEF', 1);

    expect(result).toMatchSnapshot();
  });

  it('Should return correct RGBA string for HSV color', () => {
    const result = rgba({ h: 0.5, s: 0.3, v: 0.7 }, 1);

    expect(result).toMatchSnapshot();
  });
});
