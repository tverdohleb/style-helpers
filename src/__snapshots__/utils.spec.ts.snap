// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate metric strings Should return correct REM string with default base 1`] = `"1rem "`;

exports[`Generate metric strings Should return correct RGBA string for HSV color 1`] = `"rgba(179, 125, 125, 1)"`;

exports[`Generate metric strings Should return correct RGBA string for hex color 1`] = `"rgba(171, 205, 239, 1)"`;
